File: 2023/day4.py
from functools import reduce
import re

with open("_input/day4.txt", encoding='utf8') as f:
    lines = f.read().splitlines()

part_1 = 0
card_duplicate_count = {}

def update_card_count_for_part_2(card_number, matches):
    if card_number not in card_duplicate_count:
        card_duplicate_count[card_number] = 0
    card_duplicate_count[card_number] += 1
    for _ in range(card_duplicate_count[card_number]):
        for i in range(1, matches + 1):
            if card_number + i not in card_duplicate_count:
                card_duplicate_count[card_number + i] = 0
            card_duplicate_count[card_number + i] += 1

def count_matches(arr1, arr2):
    matches = 0
    for i in arr1:
        if i in arr2:
            matches += 1
    return matches

for line in lines:
    groups = re.match('Card\s+(\d+):(.*)\|(.*)', line).groups()
    card_number = int(groups[0])
    mine = list(filter(lambda x: x, groups[1].strip().split(' ')))
    nums = list(filter(lambda x: x, groups[2].strip().split(' ')))

    matches = count_matches(mine, nums)
    part_1 += 2**matches//2

    update_card_count_for_part_2(card_number, matches)

print(f'part 1: {part_1}')
print(f'part 2: {sum(card_duplicate_count.values())}')

File: 2023/day4.go
package main

import (
	"aoc/2023/utils"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	lines := utils.GetFileLines("_input/day4.txt")
	part1 := 0
	cardDuplicateCount := make(map[int]int)
	updateCardCountForPart2 := func(cardNumber int, matches int) {
		if _, ok := cardDuplicateCount[cardNumber]; !ok {
			cardDuplicateCount[cardNumber] = 0
		}
		cardDuplicateCount[cardNumber]++
		for j := 0; j < cardDuplicateCount[cardNumber]; j++ {
			for i := 1; i <= matches; i++ {
				if _, exists := cardDuplicateCount[cardNumber+i]; !exists {
					cardDuplicateCount[cardNumber+i] = 0
				}
				cardDuplicateCount[cardNumber+i]++
			}
		}
	}

	countMatches := func(arr1, arr2 []string) int {
		matches := 0
		for _, val1 := range arr1 {
			for _, val2 := range arr2 {
				if val1 == val2 {
					matches++
					break
				}
			}
		}
		return matches
	}

	for _, line := range lines {
		re := regexp.MustCompile(`Card\s+(\d+):(.*)\|(.*)`)
		groups := re.FindStringSubmatch(line)
		cardNumber, _ := strconv.Atoi(groups[1])
		mine := strings.Fields(groups[2])
		nums := strings.Fields(groups[3])

		matches := countMatches(mine, nums)
		part1 += 1 << (matches / 2)

		updateCardCountForPart2(cardNumber, matches)
	}

	fmt.Printf("part 1: %d\n", part1)

	sumDuplicateCounts := 0
	for _, value := range cardDuplicateCount {
		sumDuplicateCounts += value
	}
	fmt.Printf("part 2: %d\n", sumDuplicateCounts)
}
